### PR TITLE
To unicode #4 - clean PR

### PR DESCRIPTION
**Pull request by miguelteixeiraa**
refs: https://github.com/ada-url/idna/pull/11, https://github.com/ada-url/idna/issues/4

All tests passing! 
I was having problems with the pre-commit hook and ended up messing with the history of the previous branch.

### CODE REVIEW

File: tests/to_unicode_tests.cpp [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+std::string read_file(std::string filename) {
+  constexpr auto read_size = std::size_t(4096);
+  auto stream = std::ifstream(filename.c_str());
+  stream.exceptions(std::ios_base::badbit);
+  auto out = std::string();
+  auto buf = std::string(read_size, '\0');
+  while (stream.read(&buf[0], read_size)) {
+    out.append(buf, 0, stream.gcount());
```
Comment (lemire) [inline tests/to_unicode_tests.cpp:30]:
```suggestion
    out.append(buf, 0, size_t(stream.gcount()));
```

Comment (lemire) [reply tests/to_unicode_tests.cpp:30]: Consider adding the above cast... it will please Visual Studio.

File: tests/to_unicode_tests.cpp [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+std::string read_file(std::string filename) {
+  constexpr auto read_size = std::size_t(4096);
+  auto stream = std::ifstream(filename.c_str());
+  stream.exceptions(std::ios_base::badbit);
+  auto out = std::string();
+  auto buf = std::string(read_size, '\0');
+  while (stream.read(&buf[0], read_size)) {
+    out.append(buf, 0, size_t(stream.gcount()));
+  }
+  out.append(buf, 0, stream.gcount());
```
Comment (anonrig) [inline tests/to_unicode_tests.cpp:32]:
```suggestion
  out.append(buf, 0, size_t(stream.gcount()));
```

**Review by anonrig — APPROVED** [review]

File: src/to_ascii.cpp [hunk 3]
```diff
@@ -22,7 +22,7 @@ bool constexpr begins_with(std::string_view view, std::string_view prefix) {
   if (view.size() < prefix.size()) {
     return false;
   }
-  return view.substr(0, prefix.size()) == prefix;
+  return view.find(prefix) == 0;
```
Comment (lemire) [inline src/to_ascii.cpp:25]: This change is almost certainly a massive downgrade as far as performance goes. Doing `return view.substr(0, prefix.size()) == prefix;` is just a single `memcmp` call, it is very cheap.

Looking for the position of the string prefix, and then checking that it is zero is conceptually more expensive. Looking at the assembly output in GCC, it is almost much worse.

Unless you have benchmarks supporting this change, I would oppose it, out of caution.

**Review by lemire — CHANGES_REQUESTED** [review]
The change made to begins_with seems detrimental. I could be wrong, but if so, I would like to see evidence that I am wrong.

File: src/to_ascii.cpp (hunk 3 above)
Comment (miguelteixeiraa) [reply src/to_ascii.cpp:25]: I'll do the bench!!

**Comment by lemire** [comment]
@miguelteixeiraa Do you recommend merging the PR at this time?

**Comment by miguelteixeiraa** [comment]
Yes, I think we are good to merge : ) !

**Comment by miguelteixeiraa** [comment]
Do you have anything in mind that you think would be interesting to change/add before merge?

**Comment by lemire** [comment]
Merged!

**Pull request merged**